### PR TITLE
[OAPIF provider] OGC API features 2 / CRS related fixes (fixes #52228)

### LINKED ISSUE #52228: Problems with the support for OGC API Features Part 2
### What is the bug or the crash?

Issue #50051 and PR #50468 added support for OGC API - Features - Part 2: Coordinate Reference Systems by Reference in the OAPIF provider in the QGIS client. Thank you for supporting this capability in QGIS! However, the implementation seems to be incomplete and not really support Part 2. I have identified the following problems:

1. From the list of CRSs of a collection, the first valid CRS identifier is selected and used as the CRS ([link to the code](https://github.com/qgis/QGIS/blob/master/src/providers/wfs/oapif/qgsoapifcollection.cpp#L301-L318)). All other CRSs in the list are ignored. As a result, unlike with WFS, the button to select a different CRS in the dialog to add a WFS/OGC API Features layer is not enabled. That is, the provider only supports the first CRS in the list, all other CRSs are not supported by the provider.

    As an side, in my view the strategy that GDAL uses is the best approach to avoid unnecessary coordinate transformations: use the `storageCrs` (and `storageCrsCoordinateEpoch`) if provided. See https://github.com/OSGeo/gdal/pull/6872.

2. I also made a test with an API where the first CRS in the list is the storage CRS, a projected CRS (EPSG:25832, ETRS89/UTM32N). Then the OAPIF provider selects EPSG:25832 for the layer, but requests to fetch features are constructed incorrectly. Here is a sample request sent be QGIS for the layer: `https://demo.ldproxy.net/vineyards/collections/vineyards/items?limit=100&bbox=1071472.15634381677955389,-1658938.48726865835487843,2375633.29574575973674655,-526230.92263901780825108&bbox-crs=http://www.opengis.net/def/crs/EPSG/0/25832`. The issues are:

    * To request coordinates in the layer CRS (EPSG:25832), the request needs a parameter `crs=http://www.opengis.net/def/crs/EPSG/0/25832`, but that is missing.
    * The `bbox-crs` parameter indicates that the coordinates in `bbox` are in EPSG:25832, but they are in a different CRS (the default CRS OGC:CRS84).

3. For APIs with a larger number of collections that support the same CRSs, there standard also allows [to declare a global CRS list once](https://docs.opengeospatial.org/is/18-058r1/18-058r1.html#_global_list_of_crs_identifiers). In the CRS list of the collection a value of `#/crs` then states that all CRSs in the global list are also supported for the collection. Such a value is ignored (considered an invalid CRS identifier). As a result, APIs using the global CRS list are not supported.

### Steps to reproduce the issue

1. Go to `Layer` > `Add Layer` > `Add WFS Layer ...`
2. Use `New` to add https://ogc-api.nrw.de/lika/v1 as a connection.
3. Press `Connect`.
4. See error: No Coordinate Reference System is detected and `Change` is disabled (the API uses a global CRS list).
    Bildschirm­foto 2023-03-14 um 19 50 16
5. Use `New` to add https://demo.ldproxy.net/daraa as a connection.
6. Press `Connect`.
7. See error: The first CRS in the list is detected (OGC:CRS84), but `Change` is disabled (all other CRSs are ignored).
    Bildschirm­foto 2023-03-14 um 19 52 57
8. Use `New` to add https://demo.ldproxy.net/vineyards as a connection.
9. Press `Connect`.
10. See error: The first CRS in the list is detected as the layer CRS (EPSG:25832), `Change` is disabled (all other CRSs are ignored).
    Bildschirm­foto 2023-03-14 um 19 55 14
11. See error: The features are accessed, but since the CRS parameter is missing (see the second problem), QGIS renders the vineyards close to Null island (red dots):
    Bildschirm­foto 2023-03-14 um 20 02 18

### Versions

QGIS-Version | 3.30.0-'s-Hertogenbosch | QGIS-Codeversion | f186b8efe0
-- | -- | -- | --
Qt-Version | 5.15.2
Python-Version | 3.9.5
GDAL-Version | 3.3.2
PROJ-Version | 8.1.1
EPSG-Registraturdatenbankversion | v10.028 (2021-07-07)
GEOS-Version | 3.9.1-CAPI-1.14.2
SQLite-Version | 3.35.2
PDAL-Version | 2.3.0
PostgreSQL-Client-Version | unknown
SpatiaLite-Version | 5.0.1
QWT-Version | 6.1.6
QScintilla2-Version | 2.11.5
BS-Version | macOS 13.2
  |   |   |  
Aktive Python-Erweiterungen
processing | 2.12.99
grassprovider | 2.12.99
db_manager | 0.1.20
MetaSearch | 0.3.6

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

### Additional context

I tried this with different implementations, with the same behavior. 

This includes APIs provided by QGIS Server. For those APIs one has to select that `OGC API - Features` has to be used as the "version". If the QGIS client selects the version, it will use WFS (and then, the client can select the CRS). This is somewhat counter-intuitive, since `Maximum` implies that `OGC API - Features` would be prefered over any of the WFS versions.

### PR DESCRIPTION
**Pull request by rouault**
[OAPIF provider] OGC API features 2 / CRS related fixes (fixes #52228)

- Takes into account storageCrs in /collection response to get the
  default crs in priority
- Append &crs=... to /items requests when the crs is not OGC:CRS84
- Make sure geometry axis order is flipped to easting/northing or
  long/lat when reading geometries from a CRS with inverted axis order
- Report the list of available CRS in the dialog box where one can
  select one
- Resolve "#/crs"

[WFS/OGCAPI] In source select dialog box, better handle list of CRS]

- Use an orderered list instead of a set, so that the first element
  is really the preferred one
- Do not trust the project CRS if the project has no layers: it is
  better to rely on the preferred CRS of the source instead

### CODE REVIEW

**Comment by rouault** [comment]
CC @domi4484

File: src/providers/wfs/qgswfssourceselect.cpp [hunk 1]
```diff
@@ -176,35 +176,33 @@ void QgsWFSSourceSelect::populateConnectionList()
...
-    ProjectCRS = projectRefSys.authid();
-  }
-
-  if ( !ProjectCRS.isEmpty() && crsSet.contains( ProjectCRS ) )
-  {
-    return ProjectCRS;
-  }
+    QgsCoordinateReferenceSystem projectRefSys = QgsProject::instance()->crs();
+    //convert to EPSG
+    QString ProjectCRS;
```
Comment (nyalldawson) [inline src/providers/wfs/qgswfssourceselect.cpp:192]:
```suggestion
    QString projectCRS;
```

File: src/providers/wfs/oapif/qgsoapifcollection.cpp [hunk 2]
```diff
@@ -298,25 +299,74 @@ bool QgsOapifCollection::deserialize( const json &j )
...
+        const auto storageCrsCoordinateEpoch = j["storageCrsCoordinateEpoch"];
+        if ( storageCrsCoordinateEpoch.is_number() )
+        {
+          crs.setCoordinateEpoch( storageCrsCoordinateEpoch.get<double>() );
+        }
+      }
+
+      layerCrsSet = true;
+      mLayerMetadata.setCrs( crs );
+      mCrsList.append( crs.authid() );
```
Comment (nyalldawson) [inline src/providers/wfs/oapif/qgsoapifcollection.cpp:323]: why are we using authid here? Does that specifications mandate use of crs from known registries?

Comment (rouault) [reply src/providers/wfs/oapif/qgsoapifcollection.cpp:323]: > Does that specifications mandate use of crs from known registries?

yes, OGCAPI Features Part 2 allows non-WGS84 CRS through http URIs, like http://www.opengis.net/def/crs/EPSG/0/2056

File: src/providers/wfs/qgswfssourceselect.cpp (hunk 1 above)
Comment (rouault) [reply src/providers/wfs/qgswfssourceselect.cpp:192]: done

**Review by domi4484 — APPROVED** [review]
Looks good to me!

**Comment by justb4** [comment]
This is  probably related: QGIS 3.28 RC of Oct 21, 2022 interworks ok with pg_featureserv (last Docker Image) , sending plain lon,lat `bbox` param without `bbox-crs`. Later QGIS versions like 3.28 LTR and 3.30 will not interwork as QGIS sends URL-format `bbox-crs` and `bbox`  in lat,lon for http://www.opengis.net/def/crs/EPSG/0/4326. This is valid but...

For example:
QGIS 3.28 RC of Oct 21, 2022 would send:
https://myserver/fs/collections/osmnl.planet_osm_roads/items?limit=5000&bbox=3.901,51.69,4.09,51.81
This is even valid in Part 2 OAPIF as defaults are assumed: 
i.e.  default `bbox-crs` is http://www.opengis.net/def/crs/OGC/1.3/CRS84 which is basically WGS84 in lon, lat as for the GeoJSON spec.

But later QGIS-versions send
 https://myserver/fs/collections/osmnl.planet_osm_roads/items?limit=5000&bbox=51.69,3.901,51.81,4.09&bbox-crs=http://www.opengis.net/def/crs/EPSG/0/4326.
Note lat,lon axis order and URL-encoded CRS.

Both requests are equivalent and comply to the OGC OAPIF (Part 2) standard. Only pg_featureserv (and possibly other OAPIF servers) will not accept the latter, only `bbox-crs` encoded as EPSG-number (but then assume lon,lat ordering). Also other parts of Part 2 are not implemented in pg_fs like the conformance clause, storageCRS, and Content-Crs HTTP header in response. That is a matter of time.

So my only request here, for the sake of interworking, is that if remote OAPIF servers do not advertise Part 2 conformance, that QGIS always sends the first request type (i.e. assuming default http://www.opengis.net/def/crs/OGC/1.3/CRS84  lon,lat order, and no `bbox-crs`) . 

See also [my comment here](https://github.com/CrunchyData/pg_featureserv/issues/6#issuecomment-1492079168). I realize many projects are in a transitional phase for OAPIF Part 2, or not even started, have various/previous versions in production, in that case it helps to be somewhat lenient. The conformance clauses in the landing pages may help.

**Comment by rouault** [comment]
> is that if remote OAPIF servers do not advertise Part 2 conformance, that QGIS always sends the first request type

Implemented per https://github.com/qgis/QGIS/pull/52366/commits/50a5c8f0370319167a4f14d12b4d8e59b7324e72

**Review by nyalldawson — APPROVED** [review]

**Comment by justb4** [comment]
Thanks @rouault !

**Comment by agiudiceandrea** [comment]
Please have a look at the issue report https://github.com/qgis/QGIS/issues/65969 "QGIS 4.0.1 - OGC API - Features - ignores storageCrs in OGC API Features collection metadata and defaults to CRS84".

**Pull request merged**